etl/src/v2/replication/apply.rs

```diff
@@ -346,101 +350,107 @@ where
+    destination: &D,
+    hook: &T,
+    max_batch_size: usize,
+) -> Result<bool, ApplyLoopError>
+where
+    D: Destination + Clone + Send + 'static,
+    T: ApplyLoopHook,
+    ApplyLoopError: From<<T as ApplyLoopHook>::Error>,
+{
+    let elapsed = state.last_batch_send_time.elapsed();
+    // `elapsed` could be zero in case current time is earlier than `last_batch_send_time`.
+    // We send the batch even in this case to make sure `last_batch_send_time` is reset to
+    // a new value and to avoid getting stuck with some events in the batch.
+    let time_to_send_batch = elapsed.is_zero() || elapsed > state.max_batch_fill_duration;
+
+    if time_to_send_batch || state.events_batch.len() >= max_batch_size || end_batch.is_some() {
+        if !state.events_batch.is_empty() {
+            // TODO: figure out if we can send a slice to the destination instead of a vec
+            // that would allow use to avoid new allocations of the `events_batch` vec and
+            // we could just call clear() on it.
+            let events_batch =
+                std::mem::replace(&mut state.events_batch, Vec::with_capacity(max_batch_size));
+            destination.write_events(events_batch).await?;
+            state.last_batch_send_time = Instant::now();
         }
 
-        // If we should break early after processing a message, we can do this in many ways:
-        // - break -> this breaks out of the loop and assumes that the last processed message was
-        //  successfully processed, so we apply all the messages up to this one.
-        // - break and discard -> this breaks out of the loop and assumes that the last processed
-        //  message was not fully processed, so we reset the `last_end_lsn` to the one of the message
-        //  before this one, so that when we apply events and notify Postgres, it's done as if the
-        //  last message was not processed.
-        //
-        // Early breaking can happen for example when a table sync worker has caught up with the apply worker
-        // but its batch contained more elements after the caught up element, in that case we don't
-        // want to process those elements, otherwise if we do, the apply worker will process them too
-        // causing duplicate data.
-        match state.early_break {
-            Some(BatchEarlyBreak::Break) => {
-                debug!("early break requested, stopping apply loop");
-                stop_apply_loop = true;
-
-                break;
-            }
-            Some(BatchEarlyBreak::BreakAndDiscard) => {
-                debug!(
-                    "early break and discard requested, resetting state and stopping apply loop"
-                );
-                *state = previous_state;
-                stop_apply_loop = true;
-
-                break;
-            }
-            None => {}
+        let mut end_loop = false;
+        if let Some(table_id) = skip_table {
+            end_loop |= !hook.skip_table(table_id).await?;
```

Comment:
Shouldn't this be outside of the sending of the batch? As far as I know we need to skip the table as soon as we get the event.

Reply:
This will be called as soon as the relation event indicates a change in schema because the `end_batch.is_some()` part in the `if time_to_send_batch || state.events_batch.len() >= max_batch_size || end_batch.is_some()` condition will be true in that case (`end_batch` will be set to `EndBatch::Exclusive`).

This again is important to call only after the batch is acked because if we set the table state to skipped before that we run the risk of missing events.